### ISSUE TITLE
Creating new Lambda Function loses race with IAM role resource created in same helm chart

### ISSUE THREAD

**Issue by rlister** [aws-controllers-k8s/community#1835]
**Describe the bug**

I have a helm chart containing IAM `Role` resource and Lambda `Function` resource.

The Function references the Role ARN in `spec.role`.

On installing the helm release, both kubernetes resources are created, but the Function resource reports:

    Message:  InvalidParameterValueException: The role defined for the function cannot be assumed by Lambda.

and the actual lambda function is not created.

If I subsequently make some innocuous change to the Function (eg change `spec.description`) and helm upgrade, the Function will sync and create a new lambda function, as expected. This works, but is not gitops-friendly.

**Steps to reproduce**

Create a helm chart with the following template and pass in values for docker `image` and AWS `AccountId`:

    ---
    apiVersion: iam.services.k8s.aws/v1alpha1
    kind: Role
    metadata:
      name: {{ .Release.Name }}
    spec:
      name: {{ .Release.Name }}
      assumeRolePolicyDocument: |
        {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              },
              "Action": "sts:AssumeRole"
            }
          ]
        }
      policies:
        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole

    ---
    apiVersion: lambda.services.k8s.aws/v1alpha1
    kind: Function
    metadata:
      name: {{ .Release.Name }}
    spec:
      name: {{ .Release.Name }}
      description: {{ .Release.Name }}
      packageType: Image
      code:
        imageURI: {{ .Values.image }}
      role: arn:aws:iam::{{ .Values.AccountId }}:role/{{ .Release.Name }}

**Expected outcome**

It appears the lambda API expects an extant IAM role at creation, and does not find one due to the eventually-consistent nature of the global IAM API. This failure is final.

I would expect the Lambda controller to wait for the resource in `role`, or possibly retry such a failure on new resource creation.

**Environment**

* Kubernetes version: 1.27
* Using EKS (yes/no), if so version? yes, 1.27
* AWS service targeted (S3, RDS, etc.) Lambda, IAM

**Comment by A-Hilaly** [comment]
@Vandita2020 maybe we can trigger a requeue when we see `The role defined for the function cannot be assumed by Lambda` during a lambda function creation?

**Comment by jessebye** [comment]
@rlister This should be fixed by https://github.com/aws-controllers-k8s/lambda-controller/pull/100 as long as you use `roleRef` instead of `role` in your function spec.

**Comment by rlister** [comment]
This makes it easier to reference the role (no more templating out the ARN), it doesn't appear to fix the race condition on initial creation: the Function still reports `The role defined for the function cannot be assumed by Lambda` and creates no underlying AWS resource until I wait for the Role, and then update the Function.

**Comment by jessebye** [comment]
@rlister When I tested this, it still has the error initially but will requeue and reconcile the Function after the Role has been created.

**Comment by rlister** [comment]
For reference, I am using the templates in the original question, but replaced `role` with:

```
  roleRef:
    from:
      name: {{ .Release.Name }}
```

**Comment by RedbackThomson** [comment]
@rlister Please double check you're using the newest version of the controller & it's CRDs. 

You should see a new condition on the Lambda function resource, before the role is created, that shows `ReferencesResolved = False`. This signals to the controller to requeue the resource, since at least one of the referenced dependencies was not ready yet. If you still see the `Role not defined` error as a `Terminal` condition, then it's not correctly using the role reference

**Comment by rlister** [comment]
@RedbackThomson I have double-checked: all controllers and CRDs are up to date. The Function resource has the new ReferencesResolved condition:

```
Status:
  Conditions:
    Last Transition Time:  2023-07-18T15:19:27Z
    Message:               Reference resolution failed
    Reason:                the referenced resource is not synced yet. resource:Role, namespace:default, name:...
    Status:                Unknown
    Type:                  ACK.ReferencesResolved
```

**Comment by rlister** [comment]
Digging deeper, I think I have this figured out: this bug _is_ fixed, but I was encountering a bug on the iam controller, reported here: #1490. My Role resource did not have a `description` field, and was thus marked incorrectly with `ACK.ResourceSynced` false.

Previously, using the `role` field with an actual iam role arn, this did not matter, but switching to `roleRef` now depends on the Role resource status, and surfaced this other bug.

TL;DR this bug is fixed, just ensure your Role has `description` field.

**Comment by a-hilaly** [comment]
This Lambda create/iam role race was fixed in 1.2.1
Can we close this @rlister . We can follow up with any separate bugs in different Issues :)

**Comment by rlister** [comment]
> Can we close this

Yes, thanks. This looks fixed to me, and the only separate bug is #1490
